Content.Server/Radio/EntitySystems/TelecomSystem.cs

```diff
@@ -1,5 +1,30 @@
+using Content.Server.Power.Components;
+using Content.Server.Radio.Components;
+using Content.Shared.Radio.Components;
+
 namespace Content.Server.Radio.EntitySystems;
 
 public sealed class TelecomSystem : EntitySystem
 {
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<RadioReceiveAttemptEvent>(OnRadioReceiveAttempt);
+    }
+
+    private void OnRadioReceiveAttempt(ref RadioReceiveAttemptEvent args)
+    {
+        if (args.Channel.LongRange)
+            return;
+        // if both source and receiver (for example) are handheld radios (but not headsets) then you don't server 
+        if (HasComp<RadioMicrophoneComponent>(args.RadioSource) && HasComp<RadioSpeakerComponent>(args.RadioReceiver))
+            return;
+        var servers = EntityQuery<TelecomServerComponent, EncryptionKeyHolderComponent, ApcPowerReceiverComponent, TransformComponent>();
+        foreach (var (_, keys, power, transform) in servers)
+        {
+            if (transform.MapID == args.MapId && power.Powered && keys.Channels.Contains(args.Channel.ID))
+                return;
```

Comment:
the issue here is that the query gets run every time someone sends a message as many times as there are headsets, which is why i suggested a cancellable sendmessage event that gets raised once and thus only runs the query once per message, no matter how many headsets are around
Also add some spaces between conditions and stuff, this looks really crampled

Reply:
oh ok